core/arch/arm/kernel/entry_a64.S

```diff
@@ -128,18 +128,17 @@ FUNC _start , :
 #if defined(CFG_CORE_SEL1_SPMC)
 	/*
 	 * With OP-TEE as SPMC at S-EL1 the SPMD (SPD_spmd) in TF-A passes
-	 * the DTB in x0, pagaeble part in x1 and the rest of the registers
-	 * are unused
+	 * the SPMC manifest in x0, the pagaeble part in x1, the system DTB
+	 * in x2 and the rest of the registers are unused.
 	 */
 	mov	x19, x1		/* Save pagable part */
-	mov	x20, x0		/* Save DT address */
```

Comment:
Are we dropping the value passed in `x0`, or is it saved somewhere else?

Reply:
We are dropping it since we currently don't look at the SPMC manifest.
If or when we need it we should save it.

Reply:
Looking more into this, I think there might be some breakage here.

- The SPMD by default was always passing the SPMC manifest (`TOS_FW_CONFIG`) in `x0`, at least on FVP. There might be some qemu specific code overriding this (I'm not sure?) but AFAICS the code for this is not platform specific [1] [2].

- OP-TEE saves this to `x20` at this current line, and later (mis)interprets this as the system dtb (or `HW_CONFIG` in TF-A's terminology) [3].

- On FVP platform the SPMD passes the Event Log buffer address in the SPMC manifest (`TOS_FW_CONFIG`) [4].

- OP-TEE is looking for the Event Log address in `HW_CONFIG`, but it's actually passed in the SPMC manifest, but OP-TEE misinterprets the SPMC manifest as the `HW_CONFIG`. So luckily we end up at the same place, the event log is correctly found, and our use case of getting the event log (on FVP, with OP-TEE as S-EL1 SPMC) is working correctly [5].

This is my understanding currently, does this make any sense?
By dropping `x0` we break this use case, so either the SPMD should pass the Event Log in `HW_CONFIG` or OP-TEE should look for the Event Log in the SPMC manifest, not sure which one is better.

[1]: https://github.com/ARM-software/arm-trusted-firmware/blob/master/common/desc_image_load.c#L249
[2]: https://trustedfirmware-a.readthedocs.io/en/latest/components/secure-partition-manager.html#spmc-boot
[3]: https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/boot.c#L1300
[4]: https://github.com/ARM-software/arm-trusted-firmware/blob/master/plat/arm/board/fvp/fvp_bl2_measured_boot.c#L212
[5]: https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/secure_partition.c#L581

Reply:
Thanks for looking into this. I'd like to define a common register usage for OP-TEE with SPMC at S-EL1. This thing with getting it right by mistake should also be resolved. We could if it helps break out all commits needed for FF-A 1.1 and take those in a separate PR.

https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/15339 should be updated to follow what we agree on. Thanks for the link to `populate_next_bl_params_config()`, I should use that one on QEMU too.

Reply:
> I'd like to define a common register usage for OP-TEE with SPMC at S-EL1.

If OP-TEE is the S-EL1 SPMC, I think the current solution on FVP (`x0`: SPMC manifest, `x1`: HW_CONFIG) sounds reasonable. Is it feasible for QEMU to align with this? I'm not familiar with that platform.

> This thing with getting it right by mistake should also be resolved.

I plan to open a PR soon (next 1-2 weeks) which should address this too.

> We could if it helps break out all commits needed for FF-A 1.1 and take those in a separate PR.

Sounds good to me, thanks.